File: packages/auth/amplify_auth_cognito_ios/ios/Classes/FlutterUpdateUserAttributeRequest.swift

/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import Foundation
import Amplify
import AmplifyPlugins
import amplify_core

struct FlutterUpdateUserAttributeRequest {
    var attribute: AuthUserAttribute!
    var options: AuthUpdateUserAttributeRequest.Options?
    
    init(dict: NSMutableDictionary) {
        self.attribute = self.createAttribute(rawAttribute: dict["attribute"] as! Dictionary<String, Any>)
        self.options = createOptions(options: dict["options"] as! Dictionary<String, Any>?)
    }
    
    func createAttribute(rawAttribute: Dictionary<String, Any>) -> AuthUserAttribute {
        let key = rawAttribute["userAttributeKey"] as! String
        let value = rawAttribute["value"] as! String
        return createAuthUserAttribute(key: key, value: value)
    }
    
    func createOptions(options: Dictionary<String, Any>?) -> AuthUpdateUserAttributeOperation.Request.Options {
      let pluginOptions = AWSUpdateUserAttributeOptions(
          metadata: options?["clientMetadata"] as? [String : String]
      )
      return AuthUpdateUserAttributeOperation.Request.Options(pluginOptions: pluginOptions)
    }
    
    static func validate(dict: NSMutableDictionary) throws {
        let validationErrorMessage = "UpdateUserAttributeRequest Request malformed."
        if (dict["attribute"] == nil || !(dict["attribute"] is Dictionary<String, Any>)) {
            throw InvalidRequestError.auth(comment: validationErrorMessage, suggestion: String(format: ErrorMessages.missingAttribute, "attribute"))
        } else {
            let attributeMap = dict["attribute"] as! Dictionary<String, Any>;
            if (attributeMap["userAttributeKey"] == nil) {
                throw InvalidRequestError.auth(comment: validationErrorMessage, suggestion: String(format: ErrorMessages.missingAttribute, "userAttributeKey"))
            } else if (attributeMap["value"] == nil) {
                throw InvalidRequestError.auth(comment: validationErrorMessage, suggestion: String(format: ErrorMessages.missingAttribute, "value"))
            } else if (!(attributeMap["value"] is String)) {
                // iOS SDK expects a string for user attr values, regardless of the configuration in cognito
                throw InvalidRequestError.auth(comment: validationErrorMessage, suggestion: "Attribute value is not a String.")
            }
        }
    }
}

File: packages/auth/amplify_auth_cognito_android/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/types/FlutterUpdateUserAttributeRequest.kt
/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

package com.amazonaws.amplify.amplify_auth_cognito.types

import androidx.annotation.NonNull
import com.amazonaws.amplify.amplify_auth_cognito.utils.createAuthUserAttribute
import com.amazonaws.amplify.amplify_auth_cognito.utils.validateUserAttribute
import com.amazonaws.amplify.amplify_core.exception.ExceptionMessages
import com.amazonaws.amplify.amplify_core.exception.InvalidRequestException
import com.amplifyframework.auth.AuthUserAttribute
import com.amplifyframework.auth.cognito.options.AWSCognitoAuthUpdateUserAttributeOptions

data class FlutterUpdateUserAttributeRequest(val map: HashMap<String, *>) {

    val attribute: AuthUserAttribute = createAttribute(map["attribute"] as HashMap<String, *>)
    val options: AWSCognitoAuthUpdateUserAttributeOptions = createOptions(map["options"] as HashMap<String, Any>?)

    private fun createAttribute(@NonNull rawAttribute: HashMap<String, *>): AuthUserAttribute {
        val value = rawAttribute["value"].toString()
        val key: String = rawAttribute["userAttributeKey"] as String
        val attribute: AuthUserAttribute = createAuthUserAttribute(key, value)
        return attribute
    }

    private fun createOptions(rawOptions: HashMap<String, *>?): AWSCognitoAuthUpdateUserAttributeOptions {
        val optionsBuilder = AWSCognitoAuthUpdateUserAttributeOptions.builder()
        if (rawOptions?.get("clientMetadata") != null) {
            optionsBuilder.metadata(rawOptions["clientMetadata"] as HashMap<String, String>)
        }
        return optionsBuilder.build()
    }

    companion object {
        private const val validationErrorMessage: String = "UpdateUserAttributeRequest Request malformed."
        fun validate(req: HashMap<String, *>?) {
            if (req == null) {
                throw InvalidRequestException(validationErrorMessage, ExceptionMessages.missingAttribute.format("request map"))
            } else if (!req.containsKey("attribute") || req["attribute"] !is HashMap<*, *>) {
                throw InvalidRequestException(validationErrorMessage, ExceptionMessages.missingAttribute.format("attribute"))
            } else {
                val attribute = req["attribute"] as HashMap<*, *>
                validateUserAttribute(attribute, validationErrorMessage)
            }
        }
    }
}
